R/prepExtendedComment.R

```diff
@@ -13,15 +13,24 @@
 #' @author Jan Philipp Dietrich
 #' @examples
 #' test <- function(a = 1) {
-#'    return(madrat:::prepExtendedComment(list(unit = "m", description = "example", package = "blub")))
-#'  }
-#'  test(a = 42)
+#'   return(madrat:::prepExtendedComment(list(unit = "m", description = "example", package = "blub")))
+#' }
+#' test(a = 42)
 #'
 prepExtendedComment <- function(x, type = "#undefined", warn = TRUE, n = 1) {
 
   # extract function call information for the parent call defined by n
   cl <- sys.call(-n)
-  f <- get(as.character(cl[[1]]), mode = "function", sys.frame(-n - 1))
+  functionCall <- as.character(cl[[1]])
+
+  # if readSource is called as madrat::readSource functionName will
+  # be in this unintuitive order c("::", "madrat", "readSource")
+  if (length(functionCall) == 3 && functionCall[[1]] == "::") {
+    f <- get(functionCall[[3]], envir = loadNamespace(functionCall[[2]]), mode = "function", sys.frame(-n - 1))
+  } else {
+    f <- get(functionCall, mode = "function", sys.frame(-n - 1))
+  }
+
```

Comment:
A bit unsafe to look for the function in some environment when `functionCall[[2]]` explicitly states it.  But this is an internal function, so OK.
But why not
```
functionCall <- tail(as.character(cl[[1]]), 1)
```
?

Reply:
> A bit unsafe to look for the function in some environment when functionCall[[2]] explicitly states it

I assume you refer to the else branch? This is the behavior we had previously which I did not see a reason to change.

```
functionCall <- tail(as.character(cl[[1]]), 1)
```
I assume you suggest to skip the if else and use this instead to then call `get(functionCall, mode = "function", sys.frame(-n - 1))`? This won't work in a scenario where `madrat::readSource` is called, unless `madrat` is also attached (in which case readSource likely won't be called with the `madrat::` prefix), because `get` would only look on the search path.

Reply:
> The `pos` argument can specify the environment in which to look for the object in any of several ways: as a positive integer (the position in the `search` list); as the character string name of an element in the search list; or as an `environment` (including using `sys.frame` to access the currently active function calls). The default of `-1` indicates the current environment of the call to `get`. The `envir` argument is an alternative way to specify an environment. 

—https://rdrr.io/r/base/get.html

The fully qualified call is
```
f <- get(functionCall[[3]], 
         envir = loadNamespace(functionCall[[2]]),
         mode = "function",
         pos = sys.frame(-n - 1))
```
Now, which argument is used, `envir` or `pos` (as they point to different things)?
<details>
  <summary>Spoiler:</summary>
  It is <code>envir</code>.
</details>

Since `prepExtendedComment()` is not exported, people calling it without attaching `madrat` are wholly responsible for their own pain `;)`.

But a safe formulation would be
```
f <- get(x = tail(functionCall, 1),
         envir = ifelse(1 == length(functionCall),
                        sys.frame(-n - 1),                # function() call
                        getNamespace(functionCall[[2]])), # pkg::function() call
         mode = "function")
```